gems/sorbet-runtime/lib/types/private/abstract/declare.rb

```diff
@@ -31,15 +31,16 @@ def self.declare_abstract(mod, type:)
         raise "You must call `abstract!` *before* defining an initialize method"
       end
 
-      # Don't need to silence warnings via without_ruby_warnings when calling
-      # define_method because of the guard above
-
-      mod.send(:define_method, :initialize) do |*args, &blk|
-        if self.class == mod
-          raise "#{mod} is declared as abstract; it cannot be instantiated"
+      initialize_check_module = Module.new do
+        define_method(:initialize) do |*args, &blk|
+          if mod.equal?(self.class)
+            raise "#{mod} is declared as abstract; it cannot be instantiated"
+          end
+          super(*args, &blk)
         end
-        super(*args, &blk)
       end
+
+      mod.prepend(initialize_check_module)
```

Comment:
The way sig works with `T::Private::ClassUtils.replace_method` does not work in the presence of prepend.
Using prepend here will have a lot of bad downstream consequences.

Instead of using prepend, what do you think about changing something in the on_method_added hook in `_methods.rb` that raises if trying to define a method named `:initialize` on a module that was marked abstract?

Reply:
I don't think we should ban defining `initialize` in an abstract class just because it would defeat the current instantiation check. That is also going to break code downstream.

I guess instead of prepend, I could wrap initialize on `mod` whenever someone define `initialize` on `mod`. It's kind of ugly, but at least it has localized perf impact. Does that sound okay?

Reply:
Also, looking at the comment about preprend for `T::Private::ClassUtils.replace_method`, for the specific case about sigs, I think getting rid of the check is fine. Sigs don't care about wrapping the first method in the `ancestors` array, it cares about wrapping the method that immediately comes after it. In order words, we are not actually trying to replace the method that was prepended, the problem cited in the comment. Thoughts? @jez